metals/src/main/scala/scala/meta/internal/metals/Command.scala

```diff
@@ -1,36 +1,102 @@
 package scala.meta.internal.metals
 
+import scala.reflect.ClassTag
+import scala.util.matching.Regex
+
 import scala.meta.internal.metals.MetalsEnrichments._
 
-import com.google.gson.JsonPrimitive
 import org.eclipse.{lsp4j => l}
 
+trait BaseCommand {
+  def id: String
+  def title: String
+  def description: String
+  def arguments: String
+
+  def toLSP(arguments: List[AnyRef]): l.Command =
+    new l.Command(title, id, arguments.asJava)
+
+  protected def isApplicableCommand(params: l.ExecuteCommandParams): Boolean = {
+    val command = Option(params.getCommand).getOrElse("")
+    command.stripPrefix("metals.") == id
+  }
+}
+
 case class Command(
     id: String,
     title: String,
     description: String,
     arguments: String = "`null`"
-) {
-  def unapply(string: String): Boolean = string == id
+) extends BaseCommand {
+  def unapply(params: l.ExecuteCommandParams): Boolean = {
+    isApplicableCommand(params)
+  }
 
-  def toLSP(arguments: List[AnyRef]): l.Command =
-    new l.Command(title, id, arguments.asJava)
 }
 
-object Argument {
+case class OpenBrowserCommand(
+    url: String,
+    title: String,
+    description: String
+) extends BaseCommand {
+  def id: String = s"browser-open-url:$url"
+  def arguments: String = "`null`"
 
-  def getAsString(obj: AnyRef): Option[String] = {
-    obj match {
-      case p: JsonPrimitive if p.isString => Option(p.getAsString())
+}
+
+object OpenBrowserCommand {
+  private val OpenBrowser: Regex = "browser-open-url:(.*)".r
+  def unapply(params: l.ExecuteCommandParams): Option[String] = {
+    val command = Option(params.getCommand).getOrElse("")
+    command match {
+      case OpenBrowser(url) => Some(url)
       case _ => None
     }
   }
+}
 
-  def getAsInt(obj: AnyRef): Option[Int] = {
-    obj match {
-      case p: JsonPrimitive if p.isNumber => Option(p.getAsInt())
-      case _ => None
+case class ParametrizedCommand[T: ClassTag](
+    id: String,
+    title: String,
+    description: String,
+    arguments: String
+) extends BaseCommand {
```

Comment:
I love the idea of `ParametrizedCommand`! 
Maybe at some point in time, we will be able to replace `gson` with some scala library that supports scala language constructs such as `Option`, sealed hierarchies and etc. This will give us great flexibility in the terms of how the command can look like.

Reply:
Unfortunately, we would need to replace LSP4J in that case, so for now I would rather leave it as is.